packages/eslint-plugin/src/rules/no-unnecessary-condition.ts

```diff
@@ -163,6 +176,20 @@ export default createRule<Options, MessageId>({
         node.test !== null &&
         node.test.type !== AST_NODE_TYPES.LogicalExpression
       ) {
+        /**
+         * Allow:
+         *   while (true) {}
+         *   for (;true;) {}
+         *   do {} while (true)
+         */
+        if (
+          allowConstantLoopConditions &&
+          isLoopStatement(node) &&
+          isBooleanLiteralType(getNodeType(node.test), true)
```

Comment:
I was reading through the [issue that introduced checkLoops](https://github.com/eslint/eslint/issues/5477) to `no-constant-condition`, and apparently `do { } while(false)` [is also an idiom some people use](https://stackoverflow.com/questions/2314066/do-whilefalse).  

Seems like a bit of a silly idiom to me; but it'd be fairly trivial to relax this check a little to allow it.

It'd be at the cost of also allowing `while (false)` and `for (;false;)` but that seems minor.

Reply:
I'm of the opinion that that's a terrible way to write code.
I don't think it's something we should support in this iteration.
I'd probably go as far as to say that more often than not a `while (false)` is either a typo, or it's "commenting" out a loop temporarily (and forgetting about it).